src/libopensc/card-idprime.c

```diff
@@ -0,0 +1,698 @@
+	num_entries = buf[0];
+	if (r < num_entries*21 + 1) {
+		r = SC_ERROR_INVALID_DATA;
+		goto done;
+	}
+	new_object.fd = 0;
+	for (i = 0; i < num_entries; i++) {
+		u8 *start = &buf[i*21+1];
+
+		/* First two bytes specify the object DF */
+		new_object.df[0] = start[0];
+		new_object.df[1] = start[1];
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "df=%s",
+			sc_dump_hex(new_object.df, sizeof(new_object.df)));
+		/* I assume this is some identification of certificate */
+		if (start[4] == 0x6B && start[5] == 0x78 && start[6] == 0x63
+				&& start[7] == 0x30) {
+			new_object.fd++;
+			/* The key reference is one bigger than the value found here for some reason */
+			new_object.key_reference = start[8] + 1;
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Found certificate with fd=%d",
+				new_object.fd);
+			idprime_add_object_to_list(&priv->pki_list, &new_object);
+		}
+	}
+	r = SC_SUCCESS;
+done:
+	if (buf)
+		free(buf);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+/*
+ * Look for a IDPrime card. If it exists, initialize our data structures
+ */
+static int idprime_find_and_initialize(sc_card_t *card, int initialize)
+{
+	int r;
+	idprime_private_data_t *priv = NULL;
+
+	/* already initialized? */
+	if (card->drv_data) {
+		return SC_SUCCESS;
+	}
+
+	r = idprime_select_index(card);
+	if (r > 0) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Index file found");
+		if (!initialize) {/* match card only */
+			return SC_SUCCESS;
+		}
+
+		priv = idprime_new_private_data();
+		if (!priv) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+		r = idprime_process_index(card, priv, r);
+		if (r == SC_SUCCESS) {
+			card->type = SC_CARD_TYPE_IDPRIME_GENERIC;
+			card->drv_data = priv;
+			return r;
+		}
+		idprime_free_private_data(priv);
+	}
+	return r;
+}
+
+static int idprime_init(sc_card_t *card)
+{
+	int r;
+	unsigned long flags;
+
+	/* TODO We should get this from the MF FCI:
+	 * 00 A4 04 00 00 (select MF returning FCI, tag 9F65 under tag A5 from GP
+	 */
+	card->max_send_size = (255 - 12);
+	card->max_recv_size = 255;
+
+	r = idprime_find_and_initialize(card, 1);
+	if (r < 0) {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_CARD);
+	}
+
+	card->name = "Gemalto IDPrime";
+	card->cla = 0x00;
+
+	/* Set up algorithm info. */
+	flags = SC_ALGORITHM_RSA_PAD_PKCS1
+		| SC_ALGORITHM_RSA_PAD_PSS
+		| SC_ALGORITHM_RSA_PAD_OAEP
+		/* SHA-1 mechanisms are not allowed in the card I have */
+		| (SC_ALGORITHM_RSA_HASH_SHA256 | SC_ALGORITHM_RSA_HASH_SHA384 | SC_ALGORITHM_RSA_HASH_SHA512)
+		| (SC_ALGORITHM_MGF1_SHA256 | SC_ALGORITHM_MGF1_SHA384 | SC_ALGORITHM_MGF1_SHA512)
+		;
+
+	_sc_card_add_rsa_alg(card, 2048, flags, 0);
+
+	card->caps |= SC_CARD_CAP_ISO7816_PIN_INFO;
+
+	LOG_FUNC_RETURN(card->ctx, 0);
+}
+
+static int idprime_match_card(sc_card_t *card)
+{
+	int i, r;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	i = _sc_match_atr(card, idprime_atrs, &card->type);
+	if (i < 0)
+		return 0;
+
+	r = idprime_find_and_initialize(card, 0);
```

Comment:
Ideally, you only modify the `card` object in `idprime_init`, i.e. you only add private data in card->init so that it definitely does get freed in card->finish. By the way, your card doesn't have a function for cleanup, where are you freeing the private data?

Reply:
It is as you write. The private data is modified only from the init() in similar manner as piv/cac drivers. In this case, I heavily depend on the card ATR and this one select command that is executed in the match is more like a safeguard.

Indeed, it could be rewritten to simplify this by calling `idprime_select_index()` directly from  the init and it would be probably nicer. Lets see